### PR TITLE
[JENKINS-66563] add a supported API for a plugin to insert jars into its classpath

### PR DESCRIPTION
**Pull request by jtnord**
Required for BouncyCastle plugin so it does not try and expose its
classes when running in a JVM in FIPS mode, but can be other use cases
as yet unknown.

Downstream demonstration PR in BouncyCastle plugin -> jenkinsci/bouncycastle-api-plugin#53

Currently WIP pending another PR that merges both this and #5698 and a PR that consumes this incremental in the bouncycastle-api plugin (and a unit test), but reviews of the API signature welcome.

I originally thought this would be best on the `Plugin` class as 
1) it is probably dangerous to allow plugin A to manipluate plugin B's classpath
2) you really want to call this before any other dependant `Plugins` have been `start`ed 

However as `Plugin` is deprecated having this on the `Plugin` class could impact the longevity of the API and also make it harder for any plugin to move away from extending `Plugin`

See [JENKINS-66563](https://issues.jenkins-ci.org/browse/JENKINS-66563).

### Proposed changelog entries

* Allow a plugin to dynamically insert a JAR file into its classpath.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: core/src/test/java/hudson/PluginWrapperTest.java
Comment (timja) [inline core/src/test/java/hudson/PluginWrapperTest.java:111]: indenting looks off for this test

File: core/src/main/java/hudson/PluginWrapper.java
Comment (timja) [inline core/src/main/java/hudson/PluginWrapper.java:386]: is there a companion fix we could do so that you do not have to extend the `Plugin` class to use this in a plugin?

**Review by timja — APPROVED** [review]

File: core/src/main/java/hudson/PluginWrapper.java
Comment (jtnord) [reply core/src/main/java/hudson/PluginWrapper.java:386]: so the Init  Milestones are a bit messy, ideally we could use those, but it is not a simple patch IIUC and would have a large potential to break things. (currently Initializers at early stages are not actually run at those stage - but either later or never (I do not recall which) - but if we supported that it could break some plugins that had some bad behaviour - especially if they expected a well formed classpath when they ran.

https://github.com/jenkinsci/jenkins/blob/ad427ecd669b243a5f6acddbe1497915ac171fee/core/src/main/java/hudson/PluginManager.java#L587 sorts out the dependencies - but strangely does not need to complete until Jenkins is fully up.

Extensions are only available after `EXTENSIONS_AUGMENTED` is attained (so initialized during that phase) which is after `PLUGINS_STARTED` so any `Plugin` (`* extends Plugin`) that depends on BC and has some static code to initialize etc would potentially not see the correct classpath anymore.

I'm happy for suggestions if we can avoid this new API and do something else instead, perhaps @jglick is more knowledable in this area may have some pointers?

File: core/src/main/java/hudson/PluginWrapper.java
Comment (jtnord) [inline core/src/main/java/hudson/PluginWrapper.java:389]:
```suggestion
     * @throws Exception if the File could not be inserted into the classpath for some reason.
     * @since TODO
```

File: core/src/test/java/hudson/PluginWrapperTest.java
Comment (jtnord) [inline core/src/test/java/hudson/PluginWrapperTest.java:121]: @basil FYI: one interesting thing I found writing this test is that the Enumeration for URL2Classloader is lazy, whereas it is not for ANTCLassLoader.
in other words if you do not count the size of `e1` before adding the new jar but delay it you end up counting the same size.  Not sure if that will cause any subtle issues in Jenkins, but something to potentially keep in mind.

File: core/src/test/java/hudson/PluginWrapperTest.java
Comment (jtnord) [reply core/src/test/java/hudson/PluginWrapperTest.java]: fixed in 543ba52325f20cf55fd9fb8906909cc48b2e7fd6

File: core/src/main/java/hudson/PluginWrapper.java
Comment (jglick) [inline core/src/main/java/hudson/PluginWrapper.java:390]: I would suggest `@Restricted(Beta.class)` so this is only offered to `bouncycastle-api`. I fear that plugin developers will otherwise abuse this as a way to load third-party libraries into their plugin from disk without bundling them (e.g. picking up an SDK JAR from some product) rather than creating a new `URLClassLoader` as they should.

**Review by jglick — COMMENTED** [review]
> suggestions if we can avoid this new API

I do not understand the requirements (how BC implements FIPS) well enough to say.

File: core/src/main/java/hudson/PluginWrapper.java
Comment (jtnord) [reply core/src/main/java/hudson/PluginWrapper.java:390]: > rather than creating a new URLClassLoader as they should

how would that work if the SDK jar(s) need to be exposed so they can actually be used?  
(contrived example: the AWS SDK becomes not open source how would all the AWS SDK plugin work, so that all the aws-* plugins could work)?  the SDK jars need to be in the plugin classpath - but the plugin can no longer be hosted by Jenkins as its dependencies are no longer open source

Comment (jtnord) [reply core/src/main/java/hudson/PluginWrapper.java:390]: the reason for not having this `@Restricted` is so that is is not "ok to change" after the bouncycastle plugin adpats to consume it, otherwise we are back to the position we are in now whereby this API is up for changing.

Comment (timja) [reply core/src/main/java/hudson/PluginWrapper.java:390]: > how would that work if the SDK jar(s) need to be exposed so they can actually be used?

Plugins that are co-operating in a scenario like this can configure themselves to use the correct classloader.

An example of this is the https://plugins.jenkins.io/database/ plugin which loads jars from an extension directory, creates an `AntClassLoader` and uses that when configuring the database driver

Comment (jtnord) [reply core/src/main/java/hudson/PluginWrapper.java:390]: right - these are 2 very different use cases.

Would clarifying the javadoc to state when it is appropriate to call this and a link to alternatives help alleviate you concerns @jglick?

Comment (jglick) [reply core/src/main/java/hudson/PluginWrapper.java:390]: > [th]is is not "ok to change" after the bouncycastle plugin [adapts] to consume it

Not sure I follow your concern. A beta API cannot be freely modified (or deleted) once published—you would need to coördinate with the plugins using it. (The intention of the beta mark is that call sites are written with the knowledge and consent of the API’s author.) It is easier to modify than an unrestricted API, which for all you know could be in use anywhere.

> clarifying the javadoc to state when it is appropriate to call this and a link to alternatives

Yes please. Without any core hacks like this, you can expose third-party libraries from a dynamic location in various ways, such as:

* split your plugin into two modules:
  * regular Jenkins plugin code, plus some interface encapsulating access to the lib via a minimal, simplified API
  * an implementation of that interface which compiles against a `provided` static reference copy of the library, and which is packaged in your plugin as a resource (_not_ in `WEB-INF/lib/*.jar`)
* upon request,
  * dynamically find some JAR(s) on the controller (or perhaps even agent)
  * find the bridge JAR in your plugin’s resources area
  * create some `URLClassLoader` loading them both, parented to the plugin `ClassLoader`
  * use reflection to load & instantiate the class of the bridge implementation, casting to the interface from the plugin

This is not a realistic option for something like the AWS SDK which offers a broad API that other plugins would want to compile against directly, without prior arrangement with the library wrapper plugin’s author; and I presume the same applies to BouncyCastle, though again I am not very familiar with its usage in Jenkins, much less the differences in the FIPS variant.

Comment (jtnord) [reply core/src/main/java/hudson/PluginWrapper.java:390]: > . A beta API cannot be freely modified (or deleted) once published

I understood [this](https://www.jenkins.io/blog/2018/04/30/using-the-beta-annotation/) differently, but if that is the case I will add that and update the javadoc 
(basically a change in this API would not be atomically updateable using Jenkins UI and Plugin Manager as you can not update the plugin until you have updated core, and if you update core then the existing plugin version would not start)

Comment (jglick) [reply core/src/main/java/hudson/PluginWrapper.java:390]: > a change in this API would not be atomically updateable using Jenkins UI and Plugin Manager

Indeed; you would need to arrange for a deprecation period, so that people updating the plugin at least every few months or so could be spared a broken combination.

Comment (jtnord) [reply core/src/main/java/hudson/PluginWrapper.java:415]: should be addressed in b480d6e4614d88d38df4a4c548fefa1f442c3c69

File: core/src/main/java/hudson/PluginWrapper.java
Comment (jglick) [inline core/src/main/java/hudson/PluginWrapper.java:411]: Not actually a clear example of the above text, but something anyway.

**Review by jglick — APPROVED** [review]

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

File: core/src/main/java/hudson/PluginWrapper.java
Comment (dwnusbaum) [inline core/src/main/java/hudson/PluginWrapper.java:418]: Typo?
```suggestion
    public void injectJarsToClasspath(File... jars) throws Exception {
```

File: core/src/main/java/hudson/PluginWrapper.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/PluginWrapper.java:429]: return if empty array is passed?

**Review by oleg-nenashev — APPROVED** [review]
LGTM with a Beta annotation. Generally we need a reference implementation for such new logic, but whatever

**Comment by jglick** [comment]
> Generally we need a reference implementation for such new logic

Rather, reference _call site_ in this case. Is there a matching `bouncycastle-api-plugin` PR using an incremental dep to demonstrate effectiveness?

**Comment by jtnord** [comment]
> Generally we need a reference implementation for such new logic
> Rather, reference call site in this case.

as per the summary yes there is a PR.  I have upodated it with this incrementatal, previous incremental publishing was broken but it was tested using SNAPSHOTs locally.

https://github.com/jenkinsci/bouncycastle-api-plugin/pull/53/files#diff-1865c0d5e42e1ea66883342138cfd766f8adb58e9695c227ba5106cb2d3c1ef3R63

**Pull request merged**
